### PR TITLE
Non reactive errors

### PR DESCRIPTION
**Pull request by martin-lysk**
Lix currently returns a reactive property called `errors`. With this one Lix has 3 kinds of error handlings:

1. Function throws
2. Function catches and adds errors to the errors property
3. Function catches and returns an errors object as function result

This PR removes the second case for now and moves error collection logic to the thin layer using the lix SDK.

In the future we will have defined errors returned in the return object of a function called. (compare 3.) for errors that the client can react on and can deal with and throw for unexpected errors like asertation errors wrong url formats or state issues (file system has changed unexpectedly).

### CODE REVIEW

**Comment by janfjohannes** [comment]
@martin-lysk  dont forget to fix the ci

Screenshot 2024-04-09 at 23 22 02

**Comment by martin-lysk** [comment]
We looked into it - we cant merge it until openRepo throws in cases like this one:
https://github.com/opral/monorepo/blob/de169bee1a85590e51bc2af6490855a613496772/lix/packages/client/src/openRepository.ts#L385
@janfjohannes shall i fix it in this repo or do you think we should merge it together with your branch?

File: inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx
Comment (janfjohannes) [inline inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx:294]: @martin-lysk  what is the impact of this? is this a precondition to mergin?

**Comment by janfjohannes** [comment]
> We looked into it - we cant merge it until openRepo throws in cases like this one:
> 
> https://github.com/opral/monorepo/blob/de169bee1a85590e51bc2af6490855a613496772/lix/packages/client/src/openRepository.ts#L385
> 
> 
> @janfjohannes shall i fix it in this repo or do you think we should merge it together with your branch?

as discussed on call. lets fix this iin open repo , i can merge the three lines manually on my side : )

File: inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx
Comment (janfjohannes) [reply inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx:294]: maybe its just a forgotten comment

File: inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx
Comment (janfjohannes) [inline inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx:315]: @martin-lysk  you moved the fuction up, which make 2 pane diffing impossible for the function, is there any reason, or could you move it back to the old position?

**Review by janfjohannes — CHANGES_REQUESTED** [review]

File: inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx
Comment (janfjohannes) [inline inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx:529]: @martin-lysk  you moved this function which makes diffing impossible in the pr, could you move it back or was there a reason?

File: inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx
Comment (janfjohannes) [inline inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx:521]: we should not do ! here. repo should be an argument from the previous function insteafd of unwrapping repo() signal again in the effect, i think

**Review by janfjohannes — CHANGES_REQUESTED** [review]

**Comment by samuelstroschein** [comment]
> @martin-lysk  In the future we will have defined errors returned in the return object of a function called.

Please elaborate. 

If you refer to exposing a "result" like type e.g. `const { data, error } = await repo.foo()`, I urge reconsideration. The JS ecosystem expects throws. Error handling in frontend frameworks for example, will not trigger anymore. While you could argue that "but frontend frameworks will get a wrapper", the problem is caused by steering away from standards. I know that error handling in JS is frustrating but we win nothing by deviating from the standard.

Supabase switching to result types https://github.com/supabase/supabase-js/issues/32 overwhelmingly downvotes after the change was merged. Counter proposals are happening https://github.com/supabase/supabase-js/issues/885 to revert the change are happening too.

In short, while `Result` is nice internally, exposing a result type externally is breaking standards and thereby compatibility.

**Comment by martin-lysk** [comment]
@samuelstroschein Jan and I concluded this a the way to go yesterday. @janfjohannes please hop in with your thoughts as well on this on.

My thoughts on this:

1. DX
Javascript is just broken when it comes to throwing errors. Why? We need a way to let the consumer know the cases to be handled. I agree on the standard part but the dx of just marking a function as `@throws` pretty poor - you can't let the user know what errors to expect.  

The user of the lib will have to always look at the documentation and keep track of changes in errors a function can produce to be able to react to them vs having a compile time information about the errors that should be handled. 

I see the point of making a "throwing" version as suggested by the followup ticket in supabase but as long as lix is used internally we should focus on making a solid error handling and go for the returned error property that will Definitly help us define the error handling and communicate changes over the team (via typsaftynes). 

2. Inter Process communication
Lix will most likely run in another process/thread/worker/deamon soon so we will have a interprocess communication going on between consumer and Excution and returning a Serializable result also including errors will be easier.

File: inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx
Comment (martin-lysk) [reply inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx:529]: it removed the repo argument and make it part of the state it self - i needed to move it at the position to have access to repo()

File: inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx
Comment (martin-lysk) [reply inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx]: Good catch, That was meant to fix the catching in openRepo - which it didn't because of its async behaviour :-( now since openRepo throws - this is was removed with the recent push

File: inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx
Comment (martin-lysk) [reply inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx:314]: it removed the repo argument and make it part of the state it self - i needed to move it at the position to have access to repo()

File: inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx
Comment (martin-lysk) [reply inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx:520]: @NiklasBuchfink I changed this to a bang (!) because it should be available at this state and i was fine with this for now - we discussed this on the call. but @janfjohannes also noticed that :) please decide if we should change that in this pr

**Comment by samuelstroschein** [comment]
@janfjohannes @martin-lysk I opened https://linear.app/opral/issue/INBOX-60/error-handling-in-lix-and-inlang-message-sdk. 

I don't understand your decision here. If we agree that compatibility issues will arise, a migration to throwing will come because users will complain as they do with supabase.

File: inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx
Comment (NiklasBuchfink) [reply inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx:520]: I think @janfjohannes is right, we can pass it as an argument as it is done in the previous function. This should eliminate the need for the bang.

File: inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx
Comment (martin-lysk) [reply inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx:314]: I just double checked we added those lines at the top of the function and removed the repo parameter.

```
const loadedRepo = repo()
if (!loadedRepo) {
	return { error: new PushException("Repo not loaded") }
}
```

File: inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx
Comment (martin-lysk) [reply inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx:530]: double checked the function it self has not changed

File: inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx
Comment (martin-lysk) [reply inlang/source-code/editor/src/pages/@host/@owner/@repository/State.tsx]: Last push should have tackled this one

File: inlang/source-code/sdk/src/migrations/maybeCreateFirstProjectId.test.ts
Comment (janfjohannes) [inline inlang/source-code/sdk/src/migrations/maybeCreateFirstProjectId.test.ts:24]: why was this test removed? seems unrelated

**Review by janfjohannes — APPROVED** [review]

File: inlang/source-code/sdk/src/migrations/maybeCreateFirstProjectId.test.ts
Comment (janfjohannes) [reply inlang/source-code/sdk/src/migrations/maybeCreateFirstProjectId.test.ts:24]: @martin-lysk  approved, just please check if this test was removed by accident, as it was still there last time i reviewed this pr

Comment (martin-lysk) [reply inlang/source-code/sdk/src/migrations/maybeCreateFirstProjectId.test.ts:24]: The behaviour of openRepository has changed it will fail if the repo does not exist, errors got to be handled before so - no need for this test anymore?

**Pull request merged**
